ui/components/auth/oss/password-validator.tsx

```diff
@@ -0,0 +1,144 @@
+"use client";
+
+import { AlertCircle, CheckCircle } from "lucide-react";
+import { useMemo } from "react";
+
+import {
+  PASSWORD_REQUIREMENTS,
+  passwordRequirementCheckers,
+} from "@/types/authFormSchema";
+
+interface PasswordRequirementsMessageProps {
+  password: string;
+  className?: string;
+}
+
+export const PasswordRequirementsMessage = ({
+  password,
+  className = "",
+}: PasswordRequirementsMessageProps) => {
+  const requirements = useMemo(
+    () => [
+      {
+        key: "minLength",
+        checker: passwordRequirementCheckers.minLength,
+        label: `At least ${PASSWORD_REQUIREMENTS.minLength} characters`,
+      },
+      {
+        key: "specialChars",
+        checker: passwordRequirementCheckers.specialChars,
+        label: "Special characters",
+      },
+      {
+        key: "uppercase",
+        checker: passwordRequirementCheckers.uppercase,
+        label: "Uppercase letters",
+      },
+      {
+        key: "lowercase",
+        checker: passwordRequirementCheckers.lowercase,
+        label: "Lowercase letters",
+      },
+      {
+        key: "numbers",
+        checker: passwordRequirementCheckers.numbers,
+        label: "Numbers",
+      },
+    ],
+    [],
+  );
+
+  const validationResults = useMemo(() => {
+    const hasPasswordInput = password.length > 0;
+
+    if (!hasPasswordInput) {
+      return null;
+    }
+
+    const results = requirements.map((req) => ({
+      ...req,
+      isMet: req.checker(password),
+    }));
+
+    const metCount = results.filter((r) => r.isMet).length;
+    const allRequirementsMet = metCount === requirements.length;
+
+    return {
+      results,
+      metCount,
+      allRequirementsMet,
+      hasPasswordInput,
+    };
+  }, [password, requirements]);
+
+  // Don't show anything if no password has been entered
+  if (!validationResults?.hasPasswordInput) {
+    return null;
+  }
+
+  const { results, allRequirementsMet } = validationResults;
```

Comment:
Here is not necessary too, you can go with a simple and cleaner approach:

```suggestion
  const hasPasswordInput = password.length > 0;

  if (!hasPasswordInput) {
    return null;
  }

  const results = REQUIREMENTS.map((req) => ({
    ...req,
    isMet: req.checker(password),
  }));

  const metCount = results.filter((r) => r.isMet).length;
  const allRequirementsMet = metCount === REQUIREMENTS.length;
```